src/Tools/dotnet-monitor/CollectionRules/Actions/GetEnvironmentVariableAction.cs

```diff
@@ -0,0 +1,90 @@
+using Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Exceptions;
+using Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Options.Actions;
+using Microsoft.Extensions.Logging;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Actions
+{
+    internal sealed partial class GetEnvironmentVariableActionFactory :
+        ICollectionRuleActionFactory<GetEnvironmentVariableOptions>
+    {
+        private readonly IServiceProvider _serviceProvider;
+        private readonly ILogger<GetEnvironmentVariableActionFactory> _logger;
+
+        public GetEnvironmentVariableActionFactory(IServiceProvider serviceProvider, ILogger<GetEnvironmentVariableActionFactory> logger)
+        {
+            _serviceProvider = serviceProvider ?? throw new ArgumentNullException(nameof(serviceProvider));
+            _logger = logger ?? throw new ArgumentNullException(nameof(serviceProvider));
+        }
+
+        public ICollectionRuleAction Create(IEndpointInfo endpointInfo, GetEnvironmentVariableOptions options)
+        {
+            if (null == options)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            ValidationContext context = new(options, _serviceProvider, items: null);
+            Validator.ValidateObject(options, context, validateAllProperties: true);
+
+            return new GetEnvironmentVariableAction(_logger, endpointInfo, options);
+        }
+
+        internal sealed partial class GetEnvironmentVariableAction :
+            CollectionRuleActionBase<GetEnvironmentVariableOptions>
+        {
+            private readonly IEndpointInfo _endpointInfo;
+            private readonly ILogger _logger;
+            private readonly GetEnvironmentVariableOptions _options;
+
+            public GetEnvironmentVariableAction(ILogger logger, IEndpointInfo endpointInfo, GetEnvironmentVariableOptions options)
+                : base(endpointInfo, options)
+            {
+                _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+                _endpointInfo = endpointInfo ?? throw new ArgumentNullException(nameof(endpointInfo));
+                _options = options ?? throw new ArgumentNullException(nameof(options));
+            }
+
+            protected override async Task<CollectionRuleActionResult> ExecuteCoreAsync(
+                TaskCompletionSource<object> startCompletionSource,
+                CancellationToken token)
+            {
+                try
+                {
+                    DiagnosticsClient client = new DiagnosticsClient(_endpointInfo.Endpoint);
+
+                    _logger.GettingEnvironmentVariable(_options.Name, _endpointInfo.ProcessId);
+                    Dictionary<string, string> envBlock = await client.GetProcessEnvironmentAsync(token);
```

Comment:
Something to think about... This call could be expensive because of the number of environment variable entries or the size of the values for each entry (think about how long PATH is on Windows). Is there a way we could cache this and use it on subsequent `GetEnvironmentVariable` action invocations within the same action list? I don't have a good suggestion here nor do I expect this to be addressed here, but just wanted to make you aware.

Reply:
This was something I thought about. Caching in a single action list does seem like a reasonable solution but it seemed like overkill for now. My guess is that this is never going to be used, if we get complaints we can fix it.